optuna/trial/_frozen.py

```diff
@@ -37,31 +49,31 @@ class FrozenTrial(object):
 
     def __init__(
         self,
-        number,  # type: int
-        state,  # type: TrialState
-        value,  # type: Optional[float]
-        datetime_start,  # type: Optional[datetime.datetime]
-        datetime_complete,  # type: Optional[datetime.datetime]
-        params,  # type: Dict[str, Any]
-        distributions,  # type: Dict[str, BaseDistribution]
-        user_attrs,  # type: Dict[str, Any]
-        system_attrs,  # type: Dict[str, Any]
-        intermediate_values,  # type: Dict[int, float]
-        trial_id,  # type: int
-    ):
-        # type: (...) -> None
-
-        self.number = number
+        number: int,
+        state: TrialState,
+        value: Optional[float],
+        datetime_start: Optional[datetime.datetime],
+        datetime_complete: Optional[datetime.datetime],
+        params: Dict[str, Any],
+        distributions: Dict[str, BaseDistribution],
+        user_attrs: Dict[str, Any],
+        system_attrs: Dict[str, Any],
+        intermediate_values: Dict[int, float],
+        trial_id: int,
+    ) -> None:
+
+        self._number = number
         self.state = state
         self.value = value
-        self.datetime_start = datetime_start
+        self._datetime_start = datetime_start
         self.datetime_complete = datetime_complete
-        self.params = params
-        self.user_attrs = user_attrs
-        self.system_attrs = system_attrs
+        self._params = params
+        self._user_attrs = user_attrs
+        self._system_attrs = system_attrs
         self.intermediate_values = intermediate_values
         self._distributions = distributions
         self._trial_id = trial_id
+        self._suggested_params = {}  # type: Dict[str, Any]
```

Comment:
The introduction of this attribute is a breaking change for `__eq__`. In this PR, this attribute is never really read from and can actually be removed. (Note that it is used in `FixedTrial` as the return value of the `params` attribute, which differs from here)

Reply:
Some attributes by the way cannot be kept consistent with both `Trial` (lazily populated) and `FrozenTrial` (set at construction). It might be okay to favor the latter as internal implementation relies on it. Those include `params`, `{user,system}_attrs`, `distributions`, etc. What do you think? Similar to how pruning works, it might be the case that we just have to document it.

Reply:
Just documenting it sounds simple at once. I think that it is great to clean the logic by making code consistent as another PR.

Reply:
Do you mean you'll include the additional documentation in this PR? I'm fine with either that or creating another one afterwards. 🙂  As for making the logic "consistent", I don't think it's possible for the reason described above. `params` for instance is populated as the user calls the `suggest_*` methods, while it's populated for a `FrozenTrial` when it's first constructed. If this `FrozenTrial` is passed to the objective function and the user assumes that it should act as a regular `Trial`, he/she might encounter unexpected behavior. This however, hopefully, is not a major concern in practice. Hope that clarifies!

Reply:
Yes, I do!  I added a note that explains `FrozenTrial`s attributes could be overwritten by re-evaluation. https://github.com/optuna/optuna/pull/1503/commits/dbbc1c343e1842eb790648d916e9f45fce249f45